Pulsar Go client library

merlimat opened the pull request:
### Motivation

Go client library implemented on top of existing C++ client lib. 

Browse the code at https://github.com/merlimat/pulsar/tree/go-client/pulsar-client-go

TODOs: 
 * ~Add tests~
 * Package C++ client lib binaries as (RPM, Deb, HomeBrew)

bruth commented on pulsar-client-go/examples/consumer-listener/consumer-listener.go:
```diff
@@ -0,0 +1,55 @@
+package main
+
+import (
+	"../../pulsar"
```
As a general rule of thumb, relatives path in Go are not ideal. Instead it should be the full import path.

merlimat replied:
Good point, should it point to fully qualified name here? like `github.com/apache/incubator-pulsar/pulsar-client-go/pulsar`

bruth replied:
Correct

bruth commented on pulsar-client-go/examples/consumer-listener/consumer-listener.go:
```diff
@@ -0,0 +1,55 @@
+		fmt.Printf("Received message  msgId: %s -- content: '%s'\n",
+			msg.MessageId(), string(msg.Payload()))
+
+		consumer.Acknowledge(msg)
```
This is partially preference, but `Ack()` would be sufficient. Go tends to prefer shorter names assuming they are not ambiguous. That said I get if you have words you are trying to maintain consistency for across clients.

merlimat replied:
`Ack()` sounds good.

bruth commented on pulsar-client-go/examples/consumer-listener/consumer-listener.go:
```diff
@@ -0,0 +1,55 @@
+		consumer.Acknowledge(msg)
+	}
+
+	consumer.Close()
```
A better pattern is to use `defer consumer.Close()` after the `if err != nil` block. (Not knowing how familiar with Go you are..) `defer` will cause the call to run after the host function exits. Although this is not the case on a panic or fatal/kill termination.

bruth replied:
Leaving this here.. but I see you used `defer` below.

merlimat replied:
Yes, makes sense to always use `defer`

bruth commented on pulsar-client-go/examples/consumer-listener/consumer-listener.go:
```diff
@@ -0,0 +1,55 @@
+		Topic("my-topic").
+		SubscriptionName("my-subscription").
+		SubscriptionType(pulsar.Shared).
+		MessageListener(channel).
```
I feel that it would be more clear if the option was `MessageChannel` rather than `Listener`.

merlimat replied:
Agree, I was using `MessageListener` since I was converting from Java interfaces, then I realized than using a channel was a better option in Go

bruth commented on pulsar-client-go/examples/consumer-listener/consumer-listener.go:
```diff
@@ -0,0 +1,55 @@
+
+	// Receive messages from channel
+	for cm := range channel {
+		consumer := cm.Consumer
```
What is the reason for initializing `consumer` within the block? Is it a different value than the parent `consumer`?

merlimat replied:
In this case we don't care about the consumer, because the message can only come from that same consumer. 

However, it might be possible to use a single channel to receive messages from multiple consumers. That's the reason the channel contains a `ConsumerMessage` struct rather than a plain `Message`

bruth replied:
Ah oh ok, the shared channel use case definitely makes why you need to do this here. Since this is an example, maybe just adding a comment noting this is the reason.

bruth commented on pulsar-client-go/examples/consumer-listener/consumer-listener.go:
```diff
@@ -0,0 +1,55 @@
+
+	channel := make(chan pulsar.ConsumerMessage)
+
+	consumer, err := client.NewConsumer().
```
The builder pattern is not very idiomatic in Go. A struct whose values can be set is simpler and generally preferred. I am less familiar with how that would impact the cgo requirement (I have not written any cgo myself). But I can imagine once `Subscribe` is called which sends the network call, any underlying cgo logic can be performed along with setting defaults.

In other words, this pattern would be ideal (also note removing the field name redundancy of `Subscription`.. names are heavily driven by context in which they are used. If something sounds redundant it should be reduced).

```go
consumer := &Consumer{
  Topic: "my-topic",
  Name: "my-subscription",
  Type: pulsar.Shared,
  Channel: channel,
}

if err := consumer.Subscribe(); err != nil {
  log.Fatal(err)
}

// ...
```

bruth replied:
Changing to a struct-based pattern over a builder obviously applies to the other APIs in this PR as well..

merlimat replied:
Yes, I was trying to use the struct approach for config but I could not find a good way to deal with integer defaults.

Example: producer configuration has a "send timeout" int option, for which a sensible default is 30 seconds. The problem is that the struct default of 0 is a valid value too, so I couldn't find a way to differentiate a legit 0 from an uninitiated option

bruth replied:
Hm, yes default zero values can be a nuisance. There are a couple patterns for this case, but they are escaping me at the moment. One more general alternative is to stick with the `New*` functions to initialize a client, consumer, producer, and reader value with the defaults preset on the struct which then allows the user to override the options they want. The `New` call could take the required options.

```go
// Constructor with required options.
c := NewConsumer(topic)
// Override default
c.Timeout = 10*time.Second

// Rest..
if err := c.Subscribe(); err != nil {
}
```

So in this set, trading the builder pattern for a simple struct that is returned that can be mutated before making the connection.

bruth commented on pulsar-client-go/examples/consumer/consumer.go:
```diff
@@ -0,0 +1,54 @@
+	}
+
+	for {
+		msg, err := consumer.Receive()
```
Is there a per-message timeout by chance? If so, then `Receive` could benefit from passing `context.Context` which is another idiom for libraries doing I/O where you may want or need to specific a deadline for the work.

merlimat replied:
`Receive()` would block (on a channel) until a message is available. there's also `ReceiveWithTimeout()`. Is that what you are referring with " per-message timeout " ?

bruth replied:
Yes. So rather than having `Receive` and `ReceiveWithTimeout`, you could have `Receive(context.Context)` and that context may have a deadline set already and handle both use cases. In the implementation you would check this:

```go
// Use timeout if deadline is set..
if d, ok := ctx.Deadline(); ok {
  timeout := d.Sub(time.Now())
  // ..
}

// Otherwise, block indefinitely.
```

merlimat replied:
Actually, I've come to conclusion to remove `Receive()` and `Next()` with timeouts. If one wants to have that, the channel listener can be used.

bruth commented on pulsar-client-go/examples/consumer-listener/consumer-listener.go:
```diff
@@ -0,0 +1,55 @@
+
+func main() {
+	client := pulsar.NewClient().
+		ServiceUrl("pulsar://localhost:6650").Build()
```
Not a hard or fast rule, but acronyms should be all caps. In some cases this can look terrible, but it is generally more idiomatic. So the option should be named `ServiceURL`. That said, is there any other kind of URL the client needs to set? If not, then `URL` should be just fine.

merlimat replied:
yes, `URL` sounds better

bruth commented on pulsar-client-go/examples/consumer/consumer.go:
```diff
@@ -0,0 +1,54 @@
+		}
+
+		fmt.Printf("Received message  msgId: %s -- content: '%s'\n",
+			msg.MessageId(), string(msg.Payload()))
```
Since this is a `Message` type, having a `MessageId()` method is redundant. `ID()` would be better (again noting the caps for acronyms).

merlimat replied:
Sure, abbreviated looks better, though I prefer `Id()` since it's not technically an acronym but an abbreviation. `Ack()` , `Id()`...

bruth commented on pulsar-client-go/examples/reader/reader.go:
```diff
@@ -0,0 +1,48 @@
+	client := pulsar.NewClient().
+		ServiceUrl("pulsar://localhost:6650").Build()
+
+	reader, err := client.NewReader().Topic("my-topic").StartFromEarliest().Create()
```
Any reason `StartFromEarliest()` is not a constant/enum like subscription type? I could imagine `-2` for earliest, `-1` for latest, and `>0` as the message id for the topic.

merlimat replied:
the `MessageId` is actually an object and not just a int/long. I was try to specify a `const` pointing to such objects, but with no success..

bruth replied:
Ah right, they are objects. Yea `const` only works with primitive types. You can use pre-defined variables in this case and check identity.

```go
var (
  EarliestMessage MessageId = newMessageId()
  LatestMessage   MessageId = newMessageId()
)
reader := &Reader{
  StartMessage: EarliestMessage,
}

// Implementation..
switch reader.StartMessage {
case EarliestMessage:
  // ..
case LatestMessage:
  // ..
case nil:
  // ..
default:
  // user defined..
}
```

bruth commented on pulsar-client-go/examples/reader/reader.go:
```diff
@@ -0,0 +1,48 @@
+	}
+
+	for {
+		msg, err := reader.ReadNext()
```
Any reason to use `ReadNext` over `Receive()` like with the consumer?

merlimat replied:
This is mostly to differentiate the API semantic difference between the consumer (push) and reader (pull).

merlimat replied:
(the underlying implementation for the delivery is 99% the same, in any case)

bruth replied:
Fair enough, maybe then just `Next()`?

bruth commented on pulsar-client-go/examples/reader/reader.go:
```diff
@@ -0,0 +1,48 @@
+
+		fmt.Printf("Received message  msgId: %s -- content: '%s'\n",
+			msg.MessageId(), string(msg.Payload()))
+	}
```
Should a reader ack a message?

merlimat replied:
No, the reader needs to be managing the "message-id" manually, outside the scope of Pulsar, so no acks are needed

bruth replied:
Ok I thought so actually for the binary protocol.. I got confused because the Websocket docs requires the reader to ack (whether that is actually required or not on the server, I am not sure).

merlimat commented:
@bruth Thanks for great comments. I'll go through them in detail

merlimat commented:
@bruth Addressed most of the comments

bruth commented on pulsar-client-go/examples/consumer/consumer.go:
```diff
@@ -0,0 +1,57 @@
+	}
+
+	defer consumer.Close()
+	defer client.Close()
```
This defer should be declared above, other the client will close before the consumer.

bruth commented on pulsar-client-go/examples/producer/producer.go:
```diff
@@ -0,0 +1,57 @@
+func main() {
+	client, err := pulsar.NewClient(pulsar.ClientOptions{
+		URL:       "pulsar://localhost:6650",
+		IoThreads: 5,
```
Similarly with `Url` this should be `IOThreads`. Again not a hard and fast rule, but is the general practice.

bruth commented on pulsar-client-go/pulsar/c_consumer.go:
```diff
@@ -0,0 +1,227 @@
+	return nil
+}
+
+func (c *consumer) AckId(msgId MessageId) error {
```
suggestion: `AckId` &rarr; `AckID`

bruth commented on pulsar-client-go/pulsar/c_consumer.go:
```diff
@@ -0,0 +1,227 @@
+	}
+}
+
+func (c *consumer) Receive() (Message, error) {
```
Supporting `context.Context` may be desirable here to only wait for a period of time (along with any other call that receives are sends on a channel). Then you can do:

```go
select {
case <-ctx.Done():
  return nil, ctx.Err()

case cm := <-c.defaultChannel:
  return cm.Message, nil
}
```

bruth replied:
It is a fairly accepted pattern. A [blog post](https://blog.golang.org/context) from 2014 describes the intent. The wiki has a [section on it](https://github.com/golang/go/wiki/CodeReviewComments#contexts). It is primarily to provide control over long-running tasks. Specifically it enables setting a deadline (via a timeout + now or an explicit future time) or as a cancelable context.

In this case, since channels are being used internally in the library, this degree of control would be useful in case a client only wants to wait so long.. or more importantly a graceful shutdown can occur on an interrupt/terminate signal.

> Won't make the method a bit more complex?

Yes has been the biggest complaint early on, but haven't heard that in a while. The past few versions of the standard library have added more methods that take context natively (such as for database queries and HTTP requests, etc.).

> Since there are more blocking calls, should all of them take a context as arg?

Yes.

bruth commented on pulsar-client-go/pulsar/c_message.go:
```diff
@@ -0,0 +1,172 @@
+	return *(*[]byte)(unsafe.Pointer(slice))
+}
+
+func (m *message) Id() MessageId {
```
`Id` to `ID`

bruth commented on pulsar-client-go/pulsar/c_message.go:
```diff
@@ -0,0 +1,172 @@
+	return newMessageId(m.ptr)
+}
+
+func (m *message) PublishTime() uint64 {
```
Generally it is more useful to return `time.Time` values rather than bare integers. It also ensures the resolution (millis vs. nanos) is not misinterpreted.

bruth commented on pulsar-client-go/pulsar/client.go:
```diff
@@ -0,0 +1,97 @@
+	LogConfFilePath string
+
+	// Configure whether to use TLS encryption on the connection (default: false)
+	EnableTls bool
```
`EnableTls` to `EnableTLS`

bruth commented on pulsar-client-go/pulsar/message.go:
```diff
@@ -0,0 +1,80 @@
+	Properties map[string]string
+
+	// Set the event time for a given message
+	EventTime uint64
```
Use `time.Time`.

bruth commented on pulsar-client-go/pulsar/producer.go:
```diff
@@ -0,0 +1,167 @@
+	Topic() string
+
+	// return the producer name which could have been assigned by the system or specified by the client
+	ProducerName() string
```
Can this just be called `Name()`?

bruth commented on pulsar-client-go/pulsar/message.go:
```diff
@@ -0,0 +1,80 @@
+
+package pulsar
+
+type MessageBuilder struct {
```
In the examples above, you renamed this to `MessageOptions`. There are a few places this needs to be updated.

bruth commented:
@merlimat Looking much better! I think going with the struct-based options is a good strategy. I left more comments, further cleaning up some things for consistency, etc.

lucperkins commented on pulsar-client-go/examples/producer/producer.go:
```diff
@@ -0,0 +1,57 @@
+	defer producer.Close()
+
+	for i := 0; i < 10; i++ {
+		err := producer.Send(pulsar.MessageBuilder{
```
This statement can be changed to:

```go
if err := producer.Send(...); err != nil {
        log.Fatal(err)
}
```

merlimat commented:
@bruth @lucperkins some more polishing and added `context.Context` arg to `Send()`, `Receive()` and `Next()` calls. Please take another look.

bruth commented:
@merlimat Looks good. I noticed that the `Ack` calls don't take a context, but they (in theory) should. Since they interface with C, I realize that doesn't translate. But do those calls take a timeout or deadline by chance? i.e. wait N milliseconds for the ack. If so, then a timeout can be derived from the context.

merlimat commented:
>  I noticed that the Ack calls don't take a context, but they (in theory) should. Since they interface with C, I realize that doesn't translate. But do those calls take a timeout or deadline by chance? i.e. wait N milliseconds for the ack. If so, then a timeout can be derived from the context.

@bruth Short answer: the ack calls in Go lib are all non-blocking with no failure option.

So, in C++ lib we have both sync and async options for "ack" operations. It was the same in Java, though we found out that there's not much need for sync acknowledgement, so we just rely on best-effort acks. The rationale: 
 * If an ack fail, for an application it's difficult to know what to do (retry the ack? that can be done internally by the library, without application intervention)
 * If the ack fail, it typically is because the consumer is disconnected from broker. In any case, the messages are going to be redelivered by broker.
 * Client library can batch acknowledgements to broker to improve efficiency. This is already done in Java and will be extended to C++ as well.

For this reasons, we have made the ack operation to immediately return success. We haven't changed the method signatures in Java to maintain compatibility, but in Go since this is a new API there's no need for sync vs async ack or even returning an error.

bruth commented:
@merlimat Interesting, thanks for the explanation.

> we just rely on best-effort acks

So in other words, the command is sent to the server *optimistically* and if it fails the observation of this failure is a redelivery. Presumably (and as you said) if there is a connection issue, then the client library will retry sending the command transparently. But once it is sent we are assuming the ack will be handled.

So yes in this case you are right that the `context.Context` won't actually influence the outcome. That said, I am still torn if whether it should be part of the signature anyway for consistency or if there is any chance in the future acks could be synchronous. I will leave this for you to decide.

merlimat commented:
@bruth We have already seen it's not useful to have sync acks (given that anyway there's no ack on the ack). Since we're already moving that API semantic in other langs already, I really don't foresee to go back to that model :)

bruth commented:
@merlimat Sounds good. I will do another pass tomorrow morning, but the code looks good to me. Maybe you can try to convince a few others in the channel to review as well.

I would like to try it out locally. What build options must be used to do the build? Include `pulsar-client-cpp/lib/c` in the `CGO_LDFLAGS`?

bruth commented:
I have a couple more comments:

- Rename `MessageBuilder` to a `ProducerMessage` since it is no longer implements a builder pattern?
- It may be useful to include the message in `SendAsync` error callback in order to correlate the error to the message.

bruth commented:
A couple more observations:

- There should be control over logging, either turning it off or provide an interface to wrap it somehow so the structure can be controlled.
- I noticed I could not create a binary that statically linked libpulsar. This could be my environment, but this was the output:

```
/usr/local/Cellar/go/1.10.1/libexec/pkg/tool/darwin_amd64/link: running clang failed: exit status 1
ld: library not found for -lcrt0.o
```

merlimat commented:
>  Rename MessageBuilder to a ProducerMessage since it is no longer implements a builder pattern?

Yes, I didn't want to call it `Message` or `MessageOptions`, `ProducerMessage` sounds good.

>  It may be useful to include the message in SendAsync error callback in order to correlate the error to the message.

Yes, mostly in C++ or Java you attach that by adding the value on the lambda. Changed it here

>  There should be control over logging, either turning it off or provide an interface to wrap it somehow so the structure can be controlled.

Good point, the log4cxx based logger was a bit overkill and not well integrated. Replaced now and allow to set logger function callback.

> I noticed I could not create a binary that statically linked libpulsar. This could be my environment, but this was the output:

This I think is more related to cgo and MacOS. I have the same but it was working on linux.

bruth commented on pulsar-client-go/pulsar/logger.go:
```diff
@@ -0,0 +1,48 @@
+		return "ERROR"
+
+	default:
+		return fmt.Sprintf("%d", int(l))
```
Not a big deal, but you can omit the `int()` wrap. I think it would better do `"UNKNOWN: %d"` or something so it is more obvious what is wrong.

bruth commented on pulsar-client-go/pulsar/logger.go:
```diff
@@ -0,0 +1,48 @@
+	case INFO:
+		return "INFO "
+	case WARN:
+		return "WARN "
```
Remove trailing spaces for WARN and INFO.

bruth commented on pulsar-client-go/pulsar/logger.go:
```diff
@@ -0,0 +1,48 @@
+	DEBUG LoggerLevel = 0;
+	INFO  LoggerLevel = 1;
+	WARN  LoggerLevel = 2;
+	ERROR LoggerLevel = 3;
```
You could tidy this up..

```go
const (
  DEBUG LoggerLevel = iota
  INFO
  WARN
  ERROR
)
```

bruth replied:
Also no semicolons.

bruth commented:
Great changes.

> This I think is more related to cgo and MacOS. I have the same but it was working on linux.

Good to know. I will be deploying on Linux so that works for me.

bruth commented:
@merlimat Just ran into this using the reader example.

```
panic: interface conversion: interface {} is nil, not *pulsar.readerCallback

goroutine 17 [running, locked to thread]:
github.com/apache/incubator-pulsar/pulsar-client-go/pulsar.pulsarReaderListenerProxy(0x7000004158f8, 0x4c00b30, 0x482a630)
	/Users/ruthb/Code/go/src/github.com/apache/incubator-pulsar/pulsar-client-go/pulsar/c_reader.go:124 +0xd7
github.com/apache/incubator-pulsar/pulsar-client-go/pulsar._cgoexpwrap_07f19b7e46c2_pulsarReaderListenerProxy(0x7000004158f8, 0x4c00b30, 0x482a630)
	_cgo_gotypes.go:1579 +0x3f
```

It looks like the code assumes a `*readerCallback` will be set and this does not use the assertion check:

```go
rc, ok := restorePointer(ctx).(*readerCallback)
if ok {
  rc.channel <- ReaderMessage{rc.reader, newMessageWrapper(message)}
}
```

merlimat commented:
@bruth I fixed the issues in reader (and consumer in latest push)

@sijie PTAL, this should be complete at this point

bruth commented:
🎉Nice work!
